### PR TITLE
Allow mixed usage of boolean and string when merging OIDC claims

### PR DESCRIPTION
**Pull request by ywangd**
Certain OPs mix usage of boolean and string for boolean type OIDC claims. For example, the same "email_verified" field is presented as boolean in IdToken, but is a string of `"true"` in the response of user info. This inconsistency results in failures when we try to merge them during authorization. 

This PR introduce a small leniency so that it will merge a boolean with a string that has value of the boolean's string representation. In another word, it will merge `true` with `"true"`, also will merge `false` with `"false"`, but nothing else.

### CODE REVIEW

**Comment by ywangd** [comment]
With #58524 and this PR, my intention is to help with the number of reports for OIDC "issues".

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java [hunk 1]
```diff
@@ -670,8 +670,18 @@ static JSONObject mergeObjects(JSONObject idToken, JSONObject userInfo) {
             } else if (value1 instanceof JSONObject) {
                 idToken.put(entry.getKey(), mergeObjects((JSONObject) value1, value2));
             } else if (value1.getClass().equals(value2.getClass()) == false) {
-                throw new IllegalStateException("Error merging ID token and userinfo claim value for claim [" + entry.getKey() + "]. " +
-                    "Cannot merge [" + value1.getClass().getName() + "] with [" + value2.getClass().getName() + "]");
+                // A special handling for certain OPs that mix the usage of true and "true"
+                // Retain value from idToken as all other primitive types
+                if (value1 instanceof Boolean && value2 instanceof String
+                    && ("true".equalsIgnoreCase((String)value2) || "false".equalsIgnoreCase((String)value2))) {
```
Comment (tvernum) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java:676]: I wonder if we should care whether they're the same logical value? We don't for other cases, but I think we should be.
Perhaps:
```suggestion
                    && String.valueOf(value1).equals(value2) ) {
```

Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java]: Updated as suggested. It's a bit inconsistent with other primitive types. But I think it fits for "minimal leniency" purpose which is I am after currently. Also, as discussed, it's worthwhile to revisit the whole merging behaviour in future.

**Review by tvernum — APPROVED** [review]
LGTM

**Review by albertzaharovits — APPROVED** [review]

**Comment by jkakavas** [comment]
This feels like the wrong kind of leniency to me. We are basically coding around a specific bug in a certain OP implementation. The email_verified is a [standard claim](https://openid.net/specs/openid-connect-core-1_0.html#Claims) defined to be boolean. 

If something, we could better handle the checks for standard claims types. 

Do we have an idea about how often this comes up (I remember 1-2 in the past year) and whether or not the offending OP is aware and unwilling to fix this ?

**Comment by ywangd** [comment]
@jkakavas Thanks for the comments. Yes the `email_verified` is a standard claim so ideally it should not require lenicency at all. I have to admit this lenicency has no theoretical base and is purely a pragmatic one.

The reason is that the aforementioned OP causes two issues for us. The first one is fixed by #58524 and the second one is fixed by this PR. Without this leniency, customers would still have issues with the OP using our [documented oidc configuration](https://www.elastic.co/guide/en/elasticsearch/reference/master/oidc-guide-authentication.html#oidc-create-realm). @albertzaharovits had recently dealt with one such issue, which was my motivation for this PR. My hope is that we would stop receiving such issues from customers, which cannot be achieved without #58524 and this PR go in together.

I haven't check with the offending OP and don't know whether they are willing to fix it or even aware of it. It is the right thing to do. But in reality, when customers are facing such issues, they would always file issues for us instead of the OP.

**Comment by jkakavas** [comment]
I guess my question is what we do when this/the next OP comes with another bug. Would we solve it again on our side? When/where do we draw the line?

When a user  comes to us with this, we should point them to open an issue at their OP, not change out implementation around it.

I disagree that we should make changes for this (the other one you mention is a bug on our side so we fixed it ) but given this is merged and I'm off, I'll pick another hill :) I just wanted to state my disagreement and the reasons for it.

**Comment by ywangd** [comment]
> I disagree that we should make changes for this (the other one you mention is a bug on our side so we fixed it ) but given this is merged and I'm off, I'll pick another hill :) I just wanted to state my disagreement and the reasons for it.
>
Noted

**Pull request merged**